Source/Scene/ModelExperimental/MetadataPipelineStage.js

```diff
@@ -66,6 +91,22 @@ MetadataPipelineStage.process = function (
   processPropertyTextures(renderResources, structuralMetadata);
 };
 
+function declareMetadataClassStructs(shaderBuilder) {
+  for (const metadataType of MetadataPipelineStage.METADATA_TYPES) {
```

Comment:
I ran into problems keeping the struct declarations in the correct order in the shader.

We don't know what types are needed until we loop through the property attributes and textures. But inside that loop (e.g., in `addPropertyAttributeProperty`), we are already adding fields to the main MetadataClass struct. These fields require the individual type structs to be already declared.

If we want to avoid unnecessary type structs, we could loop through the property attributes and textures twice: 
1. Loop over property attributes and textures, gathering a `Set` of types
2. Loop over the `Set` of types, declaring the needed `<type>MetadataClass` structs in the shader
3. Declare the main `Metadata` and `MetadataClass` structs, and the `initializeMetadata` function
4. Loop over property attributes and textures (again), adding fields to the structs and lines to `initializeMetadata`

Is this a good approach, or would looping twice be too much of a performance hit?

Reply:
I do think we'd need to make a separate loop to gather up a set of types.

For now, I'd say add the loop here for now. @j9liu and I will be looking at performance testing `ModelExperimental` soon, so we'll think about the performance implications of all these metadata loops then.

Reply:
one idea would be to store more information about types (like this set of distinct types) on the `StructuralMetadata` object, but I'd rather not store more than we need until we measure performance.

Reply:
Thanks @ptrgags! I added separate loops to gather a `Set` of actually-used types from the `propertyAttributes` and `propertyTextures`, and then declared only those type structs.
We ended up repeating a very similar sort of loop 4 times, so I added a helper function `processProperties` to clean it up a little.